aaelf64/aaelf64.rst

```diff
@@ -684,14 +685,54 @@ The defined processor-specific ``st_other`` flag values are listed below:
 
 A symbol table entry that is marked with the ``STO_AARCH64_VARIANT_PCS``
 flag set in its ``st_other`` field may be associated with a function that
-follows a variant procedure call standard with different register usage
-convention from the one defined in the base procedure call standard for
-the list of argument, caller-saved and callee-saved registers [AAPCS64_].
-The rules in the `Call and Jump relocations`_ section still apply to such functions.
+follows a variant procedure call standard under which:
+
+* the caller and callee exchange information in registers that are not set
+  aside for that purpose in the base procedure call standard [AAPCS64_]; or
+
+* the processor is guaranteed or allowed to be in a certain state on
+  entry or return from the function, beyond or in conflict with the state
+  guaranteed or allowed by the base procedure call standard.
+
+For example:
+
+* The function might take arguments in registers that are not normally
+  argument registers.
+
+* The function might return values in registers that are not normally
+  return value registers.
+
+* The function might guarantee that extra register state is preserved
+  by the call.
+
+* PSTATE on entry to the function might be different from normal.
+
+* PSTATE on return from the function might be different from normal.
+
+The following types of function must be marked with
+``STO_AARCH64_VARIANT_PCS``, although the list is not intended
+to be exhaustive:
+
+* vector PCS functions [VFABI64_]
+
+* functions that take arguments in SVE registers or return values
+  in SVE registers
+
+* streaming and streaming-compatible functions [AAPCS64_]
+
+* shared-ZA functions [AAPCS64_]
+
+The rules in the `Call and Jump relocations`_ section still apply to
+variant PCS functions.
+
 If a subroutine is called via a symbol reference that is marked with
 ``STO_AARCH64_VARIANT_PCS``, then code that runs between the calling routine
 and the called subroutine must preserve the contents of all registers except
-for IP0, IP1, and the condition code flags [AAPCS64_].
+for IP0, IP1, and the condition code flags [AAPCS64_].  It must also
+preserve the processor execution mode, such as PSTATE.SM and PSTATE.ZA.
+It is not possible to provide a definitive list of which state must be
+preserved, since the intention is to allow it to grow as the architecture
+evolves.
```

Comment:
To ask a naive question:   Do we know if there's any way in which processor execution mode was changed in such intermediate code before?

I.e. while we're calling out `PSTATE.{SM,ZA}` it seems like we're also putting a new restriction on all state in some class "processor execution mode", and TBH I don't know what that class encompasses.

My first guess for what that class of state encompasses would be `PSTATE.{SM,ZA}` based on context.
But without context I would have guessed the PSTATE field group `execution state controls` as mentioned in the Arm ARM (couldn't find what "field group" the SM and ZA would end up belonging to).

Reply:
What do you think about including something like "at the time of writing this entails `PSTATE.SM` and `PSTATE.ZA`" rather than "such as `PSTATE.SM` and `PSTATE.ZA`"?

Agreed we can't make it fool-proof, just trying to make it a bit more definitive about what we understand the situation to be right now.

Reply:
I'd rather avoid putting an upper bound on what's needed, since the point is that any current upper bound isn't guaranteed to be permanent.  I've also had pushback in the past about using “at the time of writing”, since that time is more obvious to the writer than the reader.

I think we should focus not on what the list currently is (for PSTATE or for registers), but instead on explaining why certain examples are included.  If it isn't obvious from the current text why PSTATE.SM and PSTATE.ZA are included, I can try to make it clearer.

Reply:
I can see your point -- I was hesitant about using "at the time of writing" too (though I still prefer it to the existing text).

Honestly when reading this I imagined state that I'm unaware of which might start to be relied upon for a users own special functions yet our runtime clobbers because we didn't think it was relevant.

I didn't originally pick up on the viewpoint of "if it's not mentioned by the `PCS` then it's either irrelevant to the `VARIANT_PCS` or the `PCS` has not been updated yet" at all.

That said -- this is an argument based on my lack of knowledge rather than anything else, so if you're confident the pitfall I'm concerned about is not a possibility then I won't push further than the suggestions below ;-)


Few final attempts at trying to avoid that particular misread (but if you don't like them then fair enough):

Maybe a lower bound instead of an upper bound -- could we explicitly exclude all `PSTATE` bits up to the architecture version at which SME was introduced?

Alternatively, trying to focus on the why -- maybe something like modifying the last sentence here to "the intention is to allow it to grow alongside the PCS specification as the architecture evolves"?